File: lib/calendar_widget_data.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';
import 'package:shared_preferences_android/shared_preferences_android.dart';

import 'widget_constants.dart';
import 'widget_settings.dart';

/// Snapshot error written when refresh cannot produce calendar data.
class CalendarWidgetError {
  const CalendarWidgetError({required this.code, required this.message});

  final String code;
  final String message;

  factory CalendarWidgetError.fromJson(Map<String, dynamic> json) {
    return CalendarWidgetError(
      code: json['code'] as String,
      message: json['message'] as String,
    );
  }

  Map<String, dynamic> toJson() => {'code': code, 'message': message};
}

/// One event line in the widget snapshot.
class CalendarEvent {
  const CalendarEvent({
    required this.color,
    required this.fontSize,
    required this.title,
    required this.isAllDay,
    this.time,
    this.location,
  });

  final String color;
  final int fontSize;
  final String? time;
  final String title;
  final String? location;
  final bool isAllDay;

  factory CalendarEvent.fromJson(Map<String, dynamic> json) {
    return CalendarEvent(
      color: json['color'] as String,
      fontSize: json['fontsize'] as int,
      time: json['time'] as String?,
      title: json['title'] as String,
      location: json['location'] as String?,
      isAllDay: json['isAllDay'] as bool,
    );
  }

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'color': color,
      'fontsize': fontSize,
      'title': title,
      'isAllDay': isAllDay,
    };
    if (time != null) {
      json['time'] = time;
    }
    if (location != null) {
      json['location'] = location;
    }
    return json;
  }
}

/// Day grouping (`VANDAAG`, `MORGEN`, or formatted date).
class CalendarSection {
  const CalendarSection({required this.title, required this.events});

  final String title;
  final List<CalendarEvent> events;

  factory CalendarSection.fromJson(Map<String, dynamic> json) {
    final rawEvents = json['events'] as List<dynamic>;
    return CalendarSection(
      title: json['title'] as String,
      events: rawEvents
          .map((event) => CalendarEvent.fromJson(event as Map<String, dynamic>))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
    'title': title,
    'events': events.map((event) => event.toJson()).toList(),
  };
}

/// Render snapshot stored under [WidgetConstants.KEY_CALENDAR_WIDGET_DATA].
///
/// Display values are copied from [WidgetSettings] when *building* snapshots
/// (worker, [CalendarWidgetData.error]). [fromJson] expects a complete JSON
/// contract — missing fields are a writer bug.
class CalendarWidgetData {
  const CalendarWidgetData({
    required this.schemaVersion,
    this.error,
    required this.headerDate,
    required this.headerColor,
    required this.headerFontSize,
    required this.sections,
  });

  final int schemaVersion;
  final CalendarWidgetError? error;
  final String headerDate;
  final String headerColor;
  final int headerFontSize;
  final List<CalendarSection> sections;

  bool get hasError => error != null;

  bool get isEmpty =>
      sections.isEmpty || sections.every((section) => section.events.isEmpty);

  factory CalendarWidgetData.fromJson(Map<String, dynamic> json) {
    final rawError = json['error'];
    return CalendarWidgetData(
      schemaVersion: json['schemaVersion'] as int,
      error: rawError == null
          ? null
          : CalendarWidgetError.fromJson(rawError as Map<String, dynamic>),
      headerDate: json['headerDate'] as String,
      headerColor: json['headerColor'] as String,
      headerFontSize: json['headerFontsize'] as int,
      sections: (json['sections'] as List<dynamic>)
          .map(
            (section) =>
                CalendarSection.fromJson(section as Map<String, dynamic>),
          )
          .toList(),
    );
  }

  Map<String, dynamic> toJson() => {
    'schemaVersion': schemaVersion,
    'error': error?.toJson(),
    'headerDate': headerDate,
    'headerColor': headerColor,
    'headerFontsize': headerFontSize,
    'sections': sections.map((section) => section.toJson()).toList(),
  };

  factory CalendarWidgetData.fromJsonString(String json) {
    return CalendarWidgetData.fromJson(
      jsonDecode(json) as Map<String, dynamic>,
    );
  }

  String toJsonString() => jsonEncode(toJson());

  /// Fallback snapshot when refresh cannot read calendars.
  factory CalendarWidgetData.error({
    required String code,
    required String message,
    required WidgetSettings settings,
    String headerDate = '',
  }) {
    return CalendarWidgetData(
      schemaVersion: WidgetConstants.SCHEMA_VERSION,
      error: CalendarWidgetError(code: code, message: message),
      headerDate: headerDate,
      headerColor: settings.headerColor,
      headerFontSize: settings.headerFontSize,
      sections: const [],
    );
  }

  static Future<SharedPreferencesWithCache> _prefs() {
    return SharedPreferencesWithCache.create(
      sharedPreferencesOptions: const SharedPreferencesAsyncAndroidOptions(
        backend: SharedPreferencesAndroidBackendLibrary.SharedPreferences,
        originalSharedPreferencesOptions: AndroidSharedPreferencesStoreOptions(
          fileName: WidgetConstants.PREFS_NAME,
        ),
      ),
      cacheOptions: const SharedPreferencesWithCacheOptions(
        allowList: {WidgetConstants.KEY_CALENDAR_WIDGET_DATA},
      ),
    );
  }

  /// Reads the snapshot written by [CalendarRefreshWorker] (Kotlin).
  static Future<CalendarWidgetData?> loadFromPrefs() async {
    final prefs = await _prefs();
    final json = prefs.getString(WidgetConstants.KEY_CALENDAR_WIDGET_DATA);
    if (json == null || json.isEmpty) {
      return null;
    }
    return CalendarWidgetData.fromJsonString(json);
  }

  Future<void> saveToPrefs() async {
    final prefs = await _prefs();
    await prefs.setString(
      WidgetConstants.KEY_CALENDAR_WIDGET_DATA,
      toJsonString(),
    );
  }
}

File: android/app/src/main/kotlin/nl/siwoc/calendarwidget/CalendarWidgetData.kt
package nl.siwoc.calendarwidget

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/** Snapshot error written when refresh cannot produce calendar data. */
data class CalendarWidgetError(
    val code: String,
    val message: String,
) {
    fun toJson(): JSONObject = JSONObject()
        .put("code", code)
        .put("message", message)

    companion object {
        fun fromJson(json: JSONObject): CalendarWidgetError = CalendarWidgetError(
            code = json.getString("code"),
            message = json.getString("message"),
        )
    }
}

/** One event line in the widget snapshot. */
data class CalendarEvent(
    val color: String,
    val fontSize: Int,
    val title: String,
    val isAllDay: Boolean,
    val time: String? = null,
    val location: String? = null,
) {
    fun toJson(): JSONObject {
        val json = JSONObject()
            .put("color", color)
            .put("fontsize", fontSize)
            .put("title", title)
            .put("isAllDay", isAllDay)
        time?.let { json.put("time", it) }
        location?.let { json.put("location", it) }
        return json
    }

    companion object {
        fun fromJson(json: JSONObject): CalendarEvent = CalendarEvent(
            color = json.getString("color"),
            fontSize = json.getInt("fontsize"),
            title = json.getString("title"),
            isAllDay = json.getBoolean("isAllDay"),
            time = json.optString("time").takeIf { json.has("time") },
            location = json.optString("location").takeIf { json.has("location") },
        )
    }
}

/** Day grouping (`VANDAAG`, `MORGEN`, or formatted date). */
data class CalendarSection(
    val title: String,
    val events: List<CalendarEvent>,
) {
    fun toJson(): JSONObject = JSONObject()
        .put("title", title)
        .put("events", JSONArray().apply { events.forEach { put(it.toJson()) } })

    companion object {
        fun fromJson(json: JSONObject): CalendarSection {
            val rawEvents = json.getJSONArray("events")
            val events = buildList {
                for (index in 0 until rawEvents.length()) {
                    add(CalendarEvent.fromJson(rawEvents.getJSONObject(index)))
                }
            }
            return CalendarSection(
                title = json.getString("title"),
                events = events,
            )
        }
    }
}

/**
 * Render snapshot stored under [WidgetConstants.KEY_CALENDAR_WIDGET_DATA].
 *
 * Display values are copied from [WidgetSettings] when *building* snapshots
 * (worker, [error]). [fromJson] expects a complete JSON contract — missing
 * fields are a writer bug.
 */
data class CalendarWidgetData(
    val schemaVersion: Int,
    val error: CalendarWidgetError?,
    val headerDate: String,
    val headerColor: String,
    val headerFontSize: Int,
    val sections: List<CalendarSection>,
) {
    val hasError: Boolean
        get() = error != null

    val isEmpty: Boolean
        get() = sections.isEmpty() || sections.all { it.events.isEmpty() }

    fun toJson(): JSONObject {
        val json = JSONObject()
            .put("schemaVersion", schemaVersion)
            .put("headerDate", headerDate)
            .put("headerColor", headerColor)
            .put("headerFontsize", headerFontSize)
            .put(
                "sections",
                JSONArray().apply { sections.forEach { put(it.toJson()) } },
            )
        error?.let { json.put("error", it.toJson()) } ?: json.put("error", JSONObject.NULL)
        return json
    }

    fun toJsonString(): String = toJson().toString()

    fun save(context: Context) {
        context.getSharedPreferences(WidgetConstants.PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(WidgetConstants.KEY_CALENDAR_WIDGET_DATA, toJsonString())
            .apply()
    }

    companion object {
        fun fromJson(json: JSONObject): CalendarWidgetData {
            val rawSections = json.getJSONArray("sections")
            val sections = buildList {
                for (index in 0 until rawSections.length()) {
                    add(CalendarSection.fromJson(rawSections.getJSONObject(index)))
                }
            }
            val rawError = json.opt("error")
            val error = when {
                rawError == null || rawError == JSONObject.NULL -> null
                else -> CalendarWidgetError.fromJson(rawError as JSONObject)
            }
            return CalendarWidgetData(
                schemaVersion = json.getInt("schemaVersion"),
                error = error,
                headerDate = json.getString("headerDate"),
                headerColor = json.getString("headerColor"),
                headerFontSize = json.getInt("headerFontsize"),
                sections = sections,
            )
        }

        fun fromJsonString(json: String): CalendarWidgetData =
            fromJson(JSONObject(json))

        fun load(context: Context): CalendarWidgetData? {
            val json = context.getSharedPreferences(WidgetConstants.PREFS_NAME, Context.MODE_PRIVATE)
                .getString(WidgetConstants.KEY_CALENDAR_WIDGET_DATA, null)
            if (json.isNullOrBlank()) return null
            return fromJsonString(json)
        }

        /** Fallback snapshot when refresh cannot read calendars. */
        fun error(
            code: String,
            message: String,
            settings: WidgetSettings,
            headerDate: String = "",
        ): CalendarWidgetData = CalendarWidgetData(
            schemaVersion = WidgetConstants.SCHEMA_VERSION,
            error = CalendarWidgetError(code = code, message = message),
            headerDate = headerDate,
            headerColor = settings.headerColor,
            headerFontSize = settings.headerFontSize,
            sections = emptyList(),
        )
    }
}
